### ISSUE TITLE
dump label

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14504]
Migrated from [rt.perl.org#123836](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123836) (status was 'resolved')

Searchable as RT123836$

**Comment by p5pRT** [comment]
### From @hvds

AFL \(\<http&#8203;://lcamtuf\.coredump\.cx/afl/>\) isn't handling hard loops very well at the revision I'm using\, it got stuck on a program that reduced to this&#8203;:
&nbsp;  \./miniperl \-e 'label&#8203;: dump label;'

It transpires that \[perl \#123102\] broke 'dump LABEL'\, effectively turning it into 'goto LABEL'\.

I think it needs something like the below to fix\, but I've no idea what if anything we can do to test it\.

Hugo

\-\-\- a/pp\_ctl\.c
\+\+\+ b/pp\_ctl\.c
@&#8203;@&#8203; \-3121\,8 \+3121\,7 @&#8203;@&#8203; PP\(pp\_goto\)
&nbsp;        \}
&nbsp;     \}
 
\-    else \{
\-        assert\(do\_dump\);
\+    if \(do\_dump\) \{
 \#ifdef VMS
&nbsp;        if \(\!retop\) retop = PL\_main\_start;
 \#endif

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun Feb 15 03&#8203;:24&#8203;:55 2015\, hv wrote&#8203;:
> AFL \(\<http&#8203;://lcamtuf\.coredump\.cx/afl/>\) isn't handling hard loops very
> well at the revision I'm using\, it got stuck on a program that reduced
> to this&#8203;:
>   \./miniperl \-e 'label&#8203;: dump label;'
> 
> It transpires that \[perl \#123102\] broke 'dump LABEL'\, effectively
> turning it into 'goto LABEL'\.
> 
> I think it needs something like the below to fix\, but I've no idea
> what if anything we can do to test it\.

Maybe the test code could check for an undump \(or unexec\)\, and if present\, do a dump\, undump and test the result\.

Probably limited to devel builds\.

At the very least we can check that execution stops at the dump\(\)\.

Tony

> 
> Hugo
> 
> \-\-\- a/pp\_ctl\.c
> \+\+\+ b/pp\_ctl\.c
> @&#8203;@&#8203; \-3121\,8 \+3121\,7 @&#8203;@&#8203; PP\(pp\_goto\)
>         \}
>      \}
> 
> \-    else \{
> \-        assert\(do\_dump\);
> \+    if \(do\_dump\) \{
>  \#ifdef VMS
>         if \(\!retop\) retop = PL\_main\_start;
>  \#endif

This looks right to me\.

Tony

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun Feb 15 15&#8203;:45&#8203;:40 2015\, tonyc wrote&#8203;:
> At the very least we can check that execution stops at the dump\(\)\.

Like the attached\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0001-perl-123836-basic-tests-for-dump-catches-the-issue-a.patch</summary>

```diff
From 04c22d5db20abfb4535da6856cf9518c8bc26add Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Mon, 16 Feb 2015 11:54:25 +1100
Subject: [perl #123836] basic tests for dump, catches the issue as a TODO

---
 MANIFEST    |    1 +
 t/op/dump.t |   59 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 60 insertions(+)
 create mode 100644 t/op/dump.t

diff --git a/MANIFEST b/MANIFEST
index 0427a9f..7a6ab41 100644
--- a/MANIFEST
+++ b/MANIFEST
@@ -5280,6 +5280,7 @@ t/op/die.t			See if die works
 t/op/die_unwind.t		Check die/eval early-$@ backcompat hack
 t/op/dor.t			See if defined-or (//) works
 t/op/do.t			See if subroutines work
+t/op/dump.t			See if dump works.
 t/op/each_array.t		See if array iterators work
 t/op/each.t			See if hash iterators work
 t/op/evalbytes.t		See if evalbytes operator works
diff --git a/t/op/dump.t b/t/op/dump.t
new file mode 100644
index 0000000..27b95bf
--- /dev/null
+++ b/t/op/dump.t
@@ -0,0 +1,59 @@
+#!./perl
+
+# Minimally test if dump() behaves as expected
+
+BEGIN {
+    chdir 't' if -d 't';
+    @INC = qw(. ../lib);
+    require './test.pl';
+
+    skip_all_if_miniperl();
+}
+
+use Config;
+use File::Temp qw(tempdir);
+use Cwd qw(getcwd);
+
+skip_all("only tested on devel builds")
+  unless $Config{usedevel};
+
+# there may be other operating systems where it makes sense, but
+# there are some where it isn't, so limit the platforms we test
+# this on
+skip_all("no point in dumping on $^O")
+  unless $^O =~ /^(linux|.*bsd|solaris)$/;
+
+# execute in a work directory so File::Temp can clean up core dumps
+my $tmp = tempdir(CLEANUP => 1);
+
+my $start = getcwd;
+
+chdir $tmp
+  or skip_all("Cannot chdir to work directory");
+
+plan(2);
+
+# depending on how perl is built there may be extra output after
+# the A such as "Aborted".
+
+fresh_perl_like(<<'PROG', qr/\AA(?!B\z)/, {}, "plain dump quits");
+++$|;
+print qq(A);
+dump;
+print qq(B);
+PROG
+
+{
+local $TODO = "[perl #123836]";
+fresh_perl_like(<<'PROG', qr/A(?!B\z)/, {}, "dump with label quits");
+++$|;
+print qq(A);
+dump foo;
+foo:
+print qq(B);
+PROG
+}
+
+END {
+  chdir $start;
+}
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0002-perl-123836-dump-LABEL-shouldn-t-be-goto-LABEL.patch</summary>

```diff
From 6c54573b596b889c6faa9cbf383bf53c863676a4 Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Mon, 16 Feb 2015 11:55:24 +1100
Subject: [perl #123836] dump LABEL shouldn't be goto LABEL

---
 pp_ctl.c    |    3 +--
 t/op/dump.t |    3 ---
 2 files changed, 1 insertion(+), 5 deletions(-)

diff --git a/pp_ctl.c b/pp_ctl.c
index 0b5b1a2..f7cb216 100644
--- a/pp_ctl.c
+++ b/pp_ctl.c
@@ -3121,8 +3121,7 @@ PP(pp_goto)
 	}
     }
 
-    else {
-        assert(do_dump);
+    if (do_dump) {
 #ifdef VMS
 	if (!retop) retop = PL_main_start;
 #endif
diff --git a/t/op/dump.t b/t/op/dump.t
index 27b95bf..270b58c 100644
--- a/t/op/dump.t
+++ b/t/op/dump.t
@@ -43,8 +43,6 @@ dump;
 print qq(B);
 PROG
 
-{
-local $TODO = "[perl #123836]";
 fresh_perl_like(<<'PROG', qr/A(?!B\z)/, {}, "dump with label quits");
 ++$|;
 print qq(A);
@@ -52,7 +50,6 @@ dump foo;
 foo:
 print qq(B);
 PROG
-}
 
 END {
   chdir $start;
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From @hvds

On Sun Feb 15 16&#8203;:57&#8203;:13 2015\, tonyc wrote&#8203;:
> On Sun Feb 15 15&#8203;:45&#8203;:40 2015\, tonyc wrote&#8203;:
> > At the very least we can check that execution stops at the dump\(\)\.
> 
> Like the attached\.

Looks good to me\.

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun Feb 15 18&#8203;:31&#8203;:14 2015\, hv wrote&#8203;:
> On Sun Feb 15 16&#8203;:57&#8203;:13 2015\, tonyc wrote&#8203;:
> > On Sun Feb 15 15&#8203;:45&#8203;:40 2015\, tonyc wrote&#8203;:
> > > At the very least we can check that execution stops at the dump\(\)\.
> > 
> > Like the attached\.
> 
> Looks good to me\.

Applied in 2631bbca12a9d46d3451725c8a8634183b1a9d35\, 436b3d12366a58ded59ad21978f8948b71323684\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @jhi

At least in Linux \(dromedary\) the test looks quite noisy&#8203;:

t/op/dump \.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\. sh&#8203;: line 1&#8203;: 12332 Aborted                 /home/jhi/perl/perl "\-I\.\./lib" "tmp12330A" 2>&1
sh&#8203;: line 1&#8203;: 12334 Aborted                 /home/jhi/perl/perl "\-I\.\./lib" "tmp12330A" 2>&1
ok

The test passes\.  If run explicitly&#8203;:

\./perl \-Ilib t/op/dump\.t
1\.\.2
sh&#8203;: line 1&#8203;: 27144 Aborted                 /home/jhi/perl/perl "\-I\.\./lib" "tmp27142A" 2>&1
ok 1 \- plain dump quits
sh&#8203;: line 1&#8203;: 27146 Aborted                 /home/jhi/perl/perl "\-I\.\./lib" "tmp27142A" 2>&1
ok 2 \- dump with label quits

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Feb 20\, 2015 at 09&#8203;:07&#8203;:24AM \-0800\, Jarkko Hietaniemi via RT wrote&#8203;:
> At least in Linux \(dromedary\) the test looks quite noisy&#8203;:
> 
> t/op/dump \.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\. sh&#8203;: line 1&#8203;: 12332 Aborted                 /home/jhi/perl/perl "\-I\.\./lib" "tmp12330A" 2>&1
> sh&#8203;: line 1&#8203;: 12334 Aborted                 /home/jhi/perl/perl "\-I\.\./lib" "tmp12330A" 2>&1
> ok
> 
> The test passes\.  If run explicitly&#8203;:
> 
> \./perl \-Ilib t/op/dump\.t
> 1\.\.2
> sh&#8203;: line 1&#8203;: 27144 Aborted                 /home/jhi/perl/perl "\-I\.\./lib" "tmp27142A" 2>&1
> ok 1 \- plain dump quits
> sh&#8203;: line 1&#8203;: 27146 Aborted                 /home/jhi/perl/perl "\-I\.\./lib" "tmp27142A" 2>&1
> ok 2 \- dump with label quits

Fixed \(hopefully\) with this&#8203;:

commit 49c4aee9770d41c2fd7866800ef51cfa28e02b58
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Wed Mar 18 14&#8203;:10&#8203;:33 2015 \+0000
Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
CommitDate&#8203;: Wed Mar 18 14&#8203;:13&#8203;:45 2015 \+0000

&nbsp;    t/op/dump\.t&#8203;: silence 'Aborted' messages on stderr
&nbsp;    
&nbsp;    On some platforms\, t/op/dump\.t produces extraneous \(and scary\)
&nbsp;    
&nbsp;        sh&#8203;: line 1&#8203;: 17605 Aborted \(core dumped\)
&nbsp;    
&nbsp;    style messages on stderr\.
&nbsp;    
&nbsp;    Do a fork in each test to avoid this\.

M       t/op/dump\.t


\-\- 
This is a great day for France\!
&nbsp;    \-\- Nixon at Charles De Gaulle's funeral
